In [35]:
#import packages
import pandas as pd
import numpy as np
import os

#set max columns to none to make more viewable
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [36]:
#get path
path = os.getcwd()

#read in data
receipts = pd.read_csv(os.path.join(path, 'receipts.csv'))
brands = pd.read_csv(os.path.join(path, 'brands.csv'))
users = pd.read_csv(os.path.join(path, 'users.csv'))


## Single Receipt Deep Dives

In [37]:
#look at single receipt id to get a localized view of what the data does/should look like
receipt_id = receipts.id.values[0]

receipts_dive = receipts.loc[receipts.id == receipt_id, ]
receipts_dive

primaryKey                        id  \
0    5f9c74f70a7214ad070000371  5f9c74f70a7214ad07000037   
1    5f9c74f70a7214ad070000372  5f9c74f70a7214ad07000037   
2    5f9c74f70a7214ad070000373  5f9c74f70a7214ad07000037   
3    5f9c74f70a7214ad070000374  5f9c74f70a7214ad07000037   
4    5f9c74f70a7214ad070000375  5f9c74f70a7214ad07000037   
5    5f9c74f70a7214ad070000376  5f9c74f70a7214ad07000037   
6    5f9c74f70a7214ad070000377  5f9c74f70a7214ad07000037   
7    5f9c74f70a7214ad070000378  5f9c74f70a7214ad07000037   
8    5f9c74f70a7214ad070000379  5f9c74f70a7214ad07000037   
9   5f9c74f70a7214ad0700003710  5f9c74f70a7214ad07000037   
10  5f9c74f70a7214ad0700003711  5f9c74f70a7214ad07000037   

                      userId  bonusPointsEarned  \
0   5f9c74f7c88c1415cbddb839              750.0   
1   5f9c74f7c88c1415cbddb839              750.0   
2   5f9c74f7c88c1415cbddb839              750.0   
3   5f9c74f7c88c1415cbddb839              750.0   
4   5f9c74f7c88c1415cbddb839              750.0   
5   5f9c74f7c88c1415cbddb839              750.0   
6   5f9c74f7c88c1415cbddb839              750.0   
7   5f9c74f7c88c1415cbddb839              750.0   
8   5f9c74f7c88c1415cbddb839              750.0   
9   5f9c74f7c88c1415cbddb839              750.0   
10  5f9c74f7c88c1415cbddb839              750.0   

                              bonusPointsEarnedReason           createDate  \
0   Receipt number 1 completed, bonus point schedu...  2020-10-30 13:17:59   
1   Receipt number 1 completed, bonus point schedu...  2020-10-30 13:17:59   
2   Receipt number 1 completed, bonus point schedu...  2020-10-30 13:17:59   
3   Receipt number 1 completed, bonus point schedu...  2020-10-30 13:17:59   
4   Receipt number 1 completed, bonus point schedu...  2020-10-30 13:17:59   
5   Receipt number 1 completed, bonus point schedu...  2020-10-30 13:17:59   
6   Receipt number 1 completed, bonus point schedu...  2020-10-30 13:17:59   
7   Receipt number 1 completed, bonus point schedu...  2020-10-30 13:17:59   
8   Receipt number 1 completed, bonus point schedu...  2020-10-30 13:17:59   
9   Receipt number 1 completed, bonus point schedu...  2020-10-30 13:17:59   
10  Receipt number 1 completed, bonus point schedu...  2020-10-30 13:17:59   

            dateScanned finishedDate           modifyDate  \
0   2020-10-30 13:17:59          NaN  2021-01-03 08:24:54   
1   2020-10-30 13:17:59          NaN  2021-01-03 08:24:54   
2   2020-10-30 13:17:59          NaN  2021-01-03 08:24:54   
3   2020-10-30 13:17:59          NaN  2021-01-03 08:24:54   
4   2020-10-30 13:17:59          NaN  2021-01-03 08:24:54   
5   2020-10-30 13:17:59          NaN  2021-01-03 08:24:54   
6   2020-10-30 13:17:59          NaN  2021-01-03 08:24:54   
7   2020-10-30 13:17:59          NaN  2021-01-03 08:24:54   
8   2020-10-30 13:17:59          NaN  2021-01-03 08:24:54   
9   2020-10-30 13:17:59          NaN  2021-01-03 08:24:54   
10  2020-10-30 13:17:59          NaN  2021-01-03 08:24:54   

      pointsAwardedDate  pointsEarned         purchaseDate  \
0   2020-10-30 13:18:00         750.0  2020-10-29 13:17:59   
1   2020-10-30 13:18:00         750.0  2020-10-29 13:17:59   
2   2020-10-30 13:18:00         750.0  2020-10-29 13:17:59   
3   2020-10-30 13:18:00         750.0  2020-10-29 13:17:59   
4   2020-10-30 13:18:00         750.0  2020-10-29 13:17:59   
5   2020-10-30 13:18:00         750.0  2020-10-29 13:17:59   
6   2020-10-30 13:18:00         750.0  2020-10-29 13:17:59   
7   2020-10-30 13:18:00         750.0  2020-10-29 13:17:59   
8   2020-10-30 13:18:00         750.0  2020-10-29 13:17:59   
9   2020-10-30 13:18:00         750.0  2020-10-29 13:17:59   
10  2020-10-30 13:18:00         750.0  2020-10-29 13:17:59   

    purchasedItemCount rewardsReceiptStatus  totalSpent       barcode  \
0                 11.0             REJECTED         1.0  075925306254   
1                 11.0             REJECTED         1.0           NaN   
2                 11.0             REJECTED         

Looking at the output of the above table, there are multiple issues with the data, specifically in the fields that appear to originate from the 'rewardsReceiptItemList' column in the receipts.json file. First and foremost, the purchasedItemCount field in the data says that there are 11 items on the receipt, when in reality it appears there are only two items, with the second item being duplicated 10 times. It appears as though the duplicates come from an instance where the User input the data onto the receipt. Generally, a lot of the data is unpopulated, which could present further issues for any in depth analysis.

In [38]:
#Observing receipts with multiple items that were finished
id_fin = receipts.loc[(receipts.purchasedItemCount > 1) & (receipts.rewardsReceiptStatus	== 'FINISHED'), 'id'].values[0]
receipts_fin = receipts.loc[receipts.id == id_fin, ]
receipts_fin

primaryKey                        id  \
11  5f9c74f90a7214ad070000381  5f9c74f90a7214ad07000038   
12  5f9c74f90a7214ad070000382  5f9c74f90a7214ad07000038   
13  5f9c74f90a7214ad070000383  5f9c74f90a7214ad07000038   
14  5f9c74f90a7214ad070000384  5f9c74f90a7214ad07000038   
15  5f9c74f90a7214ad070000385  5f9c74f90a7214ad07000038   

                      userId  bonusPointsEarned  \
11  5f9c74e3f1937815bd2c1d73              300.0   
12  5f9c74e3f1937815bd2c1d73              300.0   
13  5f9c74e3f1937815bd2c1d73              300.0   
14  5f9c74e3f1937815bd2c1d73              300.0   
15  5f9c74e3f1937815bd2c1d73              300.0   

                              bonusPointsEarnedReason           createDate  \
11  Receipt number 4 completed, bonus point schedu...  2020-10-30 13:18:01   
12  Receipt number 4 completed, bonus point schedu...  2020-10-30 13:18:01   
13  Receipt number 4 completed, bonus point schedu...  2020-10-30 13:18:01   
14  Receipt number 4 completed, bonus point schedu...  2020-10-30 13:18:01   
15  Receipt number 4 completed, bonus point schedu...  2020-10-30 13:18:01   

            dateScanned         finishedDate           modifyDate  \
11  2020-10-30 13:18:01  2021-01-03 08:39:55  2021-01-03 08:39:55   
12  2020-10-30 13:18:01  2021-01-03 08:39:55  2021-01-03 08:39:55   
13  2020-10-30 13:18:01  2021-01-03 08:39:55  2021-01-03 08:39:55   
14  2020-10-30 13:18:01  2021-01-03 08:39:55  2021-01-03 08:39:55   
15  2020-10-30 13:18:01  2021-01-03 08:39:55  2021-01-03 08:39:55   

      pointsAwardedDate  pointsEarned         purchaseDate  \
11  2021-01-03 08:39:55         389.2  2020-10-29 13:18:01   
12  2021-01-03 08:39:55         389.2  2020-10-29 13:18:01   
13  2021-01-03 08:39:55         389.2  2020-10-29 13:18:01   
14  2021-01-03 08:39:55         389.2  2020-10-29 13:18:01   
15  2021-01-03 08:39:55         389.2  2020-10-29 13:18:01   

    purchasedItemCount rewardsReceiptStatus  totalSpent       barcode  \
11                 6.0             FINISHED        14.0  075925306254   
12                 6.0             FINISHED        14.0  075925306254   
13                 6.0             FINISHED        14.0  034100573065   
14                 6.0             FINISHED        14.0  034100573065   
15                 6.0             FINISHED        14.0  034100573065   

                     description  finalPrice  itemPrice needsFetchReview  \
11                           NaN         1.0        1.0            False   
12                           NaN        10.0       10.0            False   
13  MILLER LITE 24 PACK 12OZ CAN         1.0        1.0            False   
14  MILLER LITE 24 PACK 12OZ CAN         1.0        1.0            False   
15  MILLER LITE 24 PACK 12OZ CAN         1.0        1.0            False   

    partnerItemId preventTargetGapPoints  quantityPurchased  \
11              1                    NaN                1.0   
12              2                   True                2.0   
13              3                   True                1.0   
14              4                   True                1.0   
15              5                   True                1.0   

    userFlaggedBarcode userFlaggedNewItem  userFlaggedPrice  \
11                 NaN                NaN               NaN   
12        7.592531e+10               True              10.0   
13        3.410057e+10               True               1.0   
14        3.410057e+10               True               1.0   
15        3.410057e+10               True               1.0   

    userFlaggedQuantity  originalMetaBriteBarcode  \
11                  NaN                       NaN   
12                  2.0              7.592531e+10   
13                  1.0              3.410057e+10   
14                  1.0              3.410057e+10   
15                  1.0                       NaN   

    originalMetaBriteDescription pointsNotAwardedReason  \
11                           NaN                    NaN   
12       

Even looking at a receipt that was successfully processed through the Fetch system, there are issues with how the receipt is processed. The item price for the items is incorrect, as three 24 packs of beer would not be $1 each. There also appear to be duplicate items, or at the very least inconsistencies with how duplicate items are processed. 

The first item, the Sargento cheese, appears twice, with one of the instances having the item price listed as $1. Additionally, one of the observations as quantity purchased as two, and the other as one. It would stand to reason that the cheese should all be on one line with a quantity purchased of three if three units were actually purchased. A similar issue is apparent with the miller light packs. 

Furthermore, one of the items on the receipt-level dataset is totalSpent, corresponding to how much the user spent; for this observation, that amount is $14. Further down the dataset, however, there is a semi-populated targetPrice field that corresponds to a much higher value, $77 for the beer specifically.

Finally, the value of total points earned for the receipt, 389.2, does not appear to correspond to the total amount of the pointsEarnedItemInfo	field, which consists of two blank values for the cheese and three values of 30 for the beer.

## Rewards Group Item Duplicates

In [39]:
#Create high level statistics showcasing some of the issues with the data at a larger scale
df_rewardsCount = receipts.\
                    groupby(['id', 'rewardsGroup']).\
                    size().reset_index(name='counts').\
                    sort_values('counts', ascending=False)

df_rewardsCount.head(20)
    

id  \
595  600f2fc80a720f0535000030   
408  600996ac0a720f05fa000134   
627  600f39c30a7214ada2000030   
409  600996ac0a720f05fa000134   
597  600f2fc80a720f0535000030   
419  60099c3c0a7214ad89000135   
415  60099c3c0a7214ad89000135   
411  600996ac0a720f05fa000134   
237  600260210a720f05f300008f   
574  600f24970a720f053500002f   
239  600260210a720f05f300008f   
625  600f39c30a7214ada2000030   
235  600260210a720f05f300008f   
413  60099c3c0a7214ad89000135   
410  600996ac0a720f05fa000134   
429  6009e72c0a720f0535000003   
405  600992f90a720f05fa000133   
221  60025cb80a720f05f300008d   
92   5ff873f10a720f052300064f   
38   5ff371030a7214ada10005a5   

                                          rewardsGroup  counts  
595       ANNIE'S HOMEGROWN MULTI-SERVING MAC & CHEESE      23  
408                           BEN AND JERRYS ICE CREAM      21  
627                           BEN AND JERRYS ICE CREAM      16  
409                   DORITOS NACHO CHEESE MULTI SERVE      15  
597                           BEN AND JERRYS ICE CREAM      15  
419                                        KNORR SIDES      15  
415                   DORITOS NACHO CHEESE MULTI SERVE      14  
411                                        KNORR SIDES      14  
237  KLEENEX TRUSTED CARE FACIAL TISSUES 120 - 179 ...      14  
574                           BEN AND JERRYS ICE CREAM      13  
239                                PEPSI 12 OZ 12 PACK      13  
625       ANNIE'S HOMEGROWN MULTI-SERVING MAC & CHEESE      13  
235                   DORITOS NACHO CHEESE MULTI SERVE      12  
413                           BEN AND JERRYS ICE CREAM      12  
410  KLEENEX TRUSTED CARE FACIAL TISSUES 120 - 179 ...      12  
429                                PEPSI 12 OZ 12 PACK      12  
405  KLEENEX TRUSTED CARE FACIAL TISSUES 120 - 179 ...      11  
221  KLEENEX TRUSTED CARE FACIAL TISSUES 120 - 179 ...      11  
92                                 MILLER LITE 24 PACK      10  
38             CAPRI SUN ROARIN' WATERS BEVERAGE DRINK      10

In [40]:
df_rewardsCount.\
    groupby('counts').\
    size().\
    reset_index(name = 'Count of Counts').\
    sort_values('counts', ascending=True)

counts  Count of Counts
0        1              493
1        2               90
2        3               19
3        4               41
4        5               57
5        6                2
6        7                3
7        8                4
8        9                9
9       10               15
10      11                2
11      12                4
12      13                3
13      14                3
14      15                3
15      16                1
16      21                1
17      23                1

The above two tables show several interesting, potential issues with the data. In the first table, we can see that some of the tables show an absurd number of repeated items, intuitively it is fair to assume a majority of people would not buy 15 or 20 packages of Ben & Jerry's ice cream in a single instance. The second table, however, show that instances of these duplicates are rare; a majority of the data contains only single instances of an item being purchased.

## Item Price Analysis

In [41]:
#Create high level statistics showcasing some of the issues with the data at a larger scale
df_itemPrice = receipts.\
                    groupby(['id', 'itemPrice']).\
                    size().reset_index(name='counts').\
                    sort_values('counts', ascending=False)

df_itemPrice.head(20)

id  itemPrice  counts
3068  600f2fc80a720f0535000030       1.00      38
3229  600f39c30a7214ada2000030       1.00      25
3271  600f39c30a7214ada2000030       3.99      20
3061  600f2fc80a720f0535000030       0.50      20
3221  600f39c30a7214ada2000030       0.50      20
2820  600f0cc70a720f053500002c       0.50      18
2909  600f24970a720f053500002f       0.50      18
3110  600f2fc80a720f0535000030       3.99      17
2936  600f24970a720f053500002f       3.00      16
1976  600a1a8d0a7214ada2000008       3.99      15
3077  600f2fc80a720f0535000030       1.80      15
3255  600f39c30a7214ada2000030       3.00      15
3245  600f39c30a7214ada2000030       2.34      14
1184  60049d9d0a720f05f3000094       2.99      14
3291  600f39c30a7214ada2000030       5.99      14
3264  600f39c30a7214ada2000030       3.49      14
3094  600f2fc80a720f0535000030       3.00      14
3084  600f2fc80a720f0535000030       2.34      14
2471  600b39f60a720f053500000e       2.34      14
2270  600b139e0a7214ada200000a       2.34      14

In [42]:
df_itemPrice.\
    groupby('counts').\
    size().\
    reset_index(name = 'Count of Counts').\
    sort_values('counts', ascending=True).\
    head(10)

counts  Count of Counts
0       1             2685
1       2              445
2       3              157
3       4              120
4       5              117
5       6               38
6       7               28
7       8               16
8       9               31
9      10               14

The above tables highlights another issue, larger issue with the quality of the data; many items on the same receipt and across many receipts costing exactly $1.00, even as we know most items do not cost $1.00 (see the Miller Light example above). This could have larger implications as to how app's points are calculated and awarded.